diff --git i/qemu/include/exec/exec-all.h w/qemu/include/exec/exec-all.h
index 68c65678..2f894b31 100644
--- i/qemu/include/exec/exec-all.h
+++ w/qemu/include/exec/exec-all.h
@@ -479,10 +479,8 @@ hwaddr memory_region_section_get_iotlb(CPUState *cpu,
 
 static inline bool uc_mem_hook_installed(struct uc_struct *uc, hwaddr paddr)
 {
-    if (HOOK_EXISTS_BOUNDED(uc, UC_HOOK_MEM_FETCH_UNMAPPED, paddr))
-        return true;
-    if (HOOK_EXISTS_BOUNDED(uc, UC_HOOK_MEM_READ_UNMAPPED, paddr))
-        return true;
+    /* *_UNMAPPED hooks only fire for unmapped addresses, so they never need writes to
+       mapped pages to stay on the notdirty slow path. */
     if (HOOK_EXISTS_BOUNDED(uc, UC_HOOK_MEM_READ, paddr))
         return true;
     if (HOOK_EXISTS_BOUNDED(uc, UC_HOOK_MEM_READ_PROT, paddr))
@@ -493,8 +491,6 @@ static inline bool uc_mem_hook_installed(struct uc_struct *uc, hwaddr paddr)
         return true;
     if (HOOK_EXISTS_BOUNDED(uc, UC_HOOK_MEM_WRITE, paddr))
         return true;
-    if (HOOK_EXISTS_BOUNDED(uc, UC_HOOK_MEM_WRITE_UNMAPPED, paddr))
-        return true;
     if (HOOK_EXISTS_BOUNDED(uc, UC_HOOK_MEM_WRITE_PROT, paddr))
         return true;
     return false;
